File: ios/HangulEngine/Sources/HangulEngine/JamoSequence.swift
public enum JamoDecompositionError: Error, Equatable, Sendable {
  case emptyTarget
  case unsupportedCharacter(Character, offset: Int)
}

public enum JamoDecomposer {
  /// Converts composed Hangul into the exact two-beolsik key-jamo sequence.
  public static func keySequence(for target: String) throws -> [Character] {
    guard !target.isEmpty else { throw JamoDecompositionError.emptyTarget }

    var result: [Character] = []
    for (offset, character) in target.enumerated() {
      if let scalar = onlyScalar(in: character), (0xAC00...0xD7A3).contains(Int(scalar.value)) {
        let syllableIndex = Int(scalar.value) - 0xAC00
        let leading = HangulTables.leading[syllableIndex / (21 * 28)]
        let medial = HangulTables.medial[(syllableIndex % (21 * 28)) / 28]
        let trailing = HangulTables.trailing[syllableIndex % 28]
        result.append(leading)
        appendExpanded(medial, splitTable: HangulTables.splitMedial, to: &result)
        if let trailing {
          appendExpanded(trailing, splitTable: HangulTables.splitTrailing, to: &result)
        }
      } else if HangulTables.leadingIndex[character] != nil {
        result.append(character)
      } else if HangulTables.medialIndex[character] != nil {
        appendExpanded(character, splitTable: HangulTables.splitMedial, to: &result)
      } else if let split = HangulTables.splitTrailing[character] {
        result.append(contentsOf: [split.first, split.second])
      } else if isAllowedLiteral(character) {
        result.append(character)
      } else {
        throw JamoDecompositionError.unsupportedCharacter(character, offset: offset)
      }
    }
    return result
  }

  public static func isShiftJamo(_ character: Character) -> Bool {
    HangulTables.shiftedJamo.contains(character)
  }

  public static func containsHangul(in text: String) -> Bool {
    text.contains { character in
      if HangulTables.leadingIndex[character] != nil || HangulTables.medialIndex[character] != nil {
        return true
      }
      return onlyScalar(in: character).map { (0xAC00...0xD7A3).contains(Int($0.value)) } ?? false
    }
  }

  private static func appendExpanded(
    _ jamo: Character,
    splitTable: [Character: JamoPair],
    to result: inout [Character]
  ) {
    if let split = splitTable[jamo] {
      result.append(contentsOf: [split.first, split.second])
    } else {
      result.append(jamo)
    }
  }

  private static func isAllowedLiteral(_ character: Character) -> Bool {
    if HangulTables.allowedLiteralPunctuation.contains(character) { return true }
    return character.unicodeScalars.allSatisfy { scalar in
      scalar.properties.isWhitespace || scalar.properties.numericType != nil
    }
  }

  private static func onlyScalar(in character: Character) -> UnicodeScalar? {
    character.unicodeScalars.count == 1 ? character.unicodeScalars.first : nil
  }
}

public struct JamoJudgeState: Equatable, Sendable {
  public let target: String
  public let expectedSequence: [Character]
  public private(set) var currentIndex: Int
  public private(set) var correctCount: Int
  public private(set) var errorCount: Int

  public init(target: String) throws {
    self.target = target
    expectedSequence = try JamoDecomposer.keySequence(for: target)
    currentIndex = 0
    correctCount = 0
    errorCount = 0
  }

  public var expectedNext: Character? {
    currentIndex < expectedSequence.count ? expectedSequence[currentIndex] : nil
  }

  public var isComplete: Bool { currentIndex == expectedSequence.count }

  public var progress: Double {
    expectedSequence.isEmpty ? 1 : Double(currentIndex) / Double(expectedSequence.count)
  }

  fileprivate mutating func recordCorrect() {
    currentIndex += 1
    correctCount += 1
  }

  fileprivate mutating func recordError() {
    errorCount += 1
  }
}

public enum JamoJudgeResult: Equatable, Sendable {
  case correct(completed: Bool)
  case incorrect(expected: Character)
  case alreadyComplete
}

public enum JamoSequenceJudge {
  /// Compares one logical jamo event. Shift+jamo arrives as its resulting jamo and counts once.
  public static func evaluate(
    _ input: Character,
    state: JamoJudgeState
  ) -> (state: JamoJudgeState, result: JamoJudgeResult) {
    var next = state
    guard let expected = next.expectedNext else {
      return (next, .alreadyComplete)
    }
    guard input == expected else {
      next.recordError()
      return (next, .incorrect(expected: expected))
    }
    next.recordCorrect()
    return (next, .correct(completed: next.isComplete))
  }
}

File: android/core/hangul/src/main/kotlin/app/piyokey/core/hangul/JamoSequence.kt
package app.piyokey.core.hangul

sealed class JamoDecompositionException(message: String) : IllegalArgumentException(message) {
  data object EmptyTarget : JamoDecompositionException("Target must not be empty")

  data class UnsupportedCharacter(
    val character: Char,
    val offset: Int,
  ) : JamoDecompositionException("Unsupported character at $offset: $character")
}

object JamoDecomposer {
  /** Converts composed Hangul into the exact two-beolsik key-jamo sequence. */
  fun keySequenceFor(target: String): List<Char> {
    if (target.isEmpty()) throw JamoDecompositionException.EmptyTarget

    return buildList {
      var offset = 0
      while (offset < target.length) {
        val character = target[offset]
        if (
          character.isHighSurrogate() &&
          offset + 1 < target.length &&
          target[offset + 1].isLowSurrogate()
        ) {
          val lowSurrogate = target[offset + 1]
          val codePoint = Character.toCodePoint(character, lowSurrogate)
          if (isAllowedSupplementaryLiteral(codePoint)) {
            add(character)
            add(lowSurrogate)
            offset += 2
            continue
          }
        }
        when {
          character.code in 0xAC00..0xD7A3 -> {
            val syllableIndex = character.code - 0xAC00
            val leading = HangulTables.leading[syllableIndex / (21 * 28)]
            val medial = HangulTables.medial[(syllableIndex % (21 * 28)) / 28]
            val trailing = HangulTables.trailing[syllableIndex % 28]
            add(leading)
            appendExpanded(medial, HangulTables.splitMedial)
            trailing?.let { appendExpanded(it, HangulTables.splitTrailing) }
          }
          character in HangulTables.leadingIndex -> add(character)
          character in HangulTables.medialIndex -> {
            appendExpanded(character, HangulTables.splitMedial)
          }
          character in HangulTables.splitTrailing -> {
            val split = HangulTables.splitTrailing.getValue(character)
            add(split.first)
            add(split.second)
          }
          isAllowedLiteral(character) -> add(character)
          else -> throw JamoDecompositionException.UnsupportedCharacter(character, offset)
        }
        offset += 1
      }
    }
  }

  fun isShiftJamo(character: Char): Boolean = character in HangulTables.shiftedJamo

  fun containsHangul(text: String): Boolean = text.any { character ->
    character in HangulTables.leadingIndex ||
      character in HangulTables.medialIndex ||
      character.code in 0xAC00..0xD7A3
  }

  private fun MutableList<Char>.appendExpanded(
    jamo: Char,
    splitTable: Map<Char, JamoPair>,
  ) {
    val split = splitTable[jamo]
    if (split == null) {
      add(jamo)
    } else {
      add(split.first)
      add(split.second)
    }
  }

  private fun isAllowedLiteral(character: Char): Boolean {
    if (character in HangulTables.allowedLiteralPunctuation) return true
    return character.isWhitespace() || when (Character.getType(character)) {
      Character.DECIMAL_DIGIT_NUMBER.toInt(),
      Character.LETTER_NUMBER.toInt(),
      Character.OTHER_NUMBER.toInt(),
      -> true
      else -> false
    }
  }

  private fun isAllowedSupplementaryLiteral(codePoint: Int): Boolean =
    Character.isWhitespace(codePoint) || when (Character.getType(codePoint)) {
      Character.DECIMAL_DIGIT_NUMBER.toInt(),
      Character.LETTER_NUMBER.toInt(),
      Character.OTHER_NUMBER.toInt(),
      -> true
      else -> false
    }
}

@ConsistentCopyVisibility
data class JamoJudgeState private constructor(
  val target: String,
  val expectedSequence: List<Char>,
  val currentIndex: Int,
  val correctCount: Int,
  val errorCount: Int,
) {
  val expectedNext: Char?
    get() = expectedSequence.getOrNull(currentIndex)

  val isComplete: Boolean
    get() = currentIndex == expectedSequence.size

  val progress: Double
    get() = currentIndex.toDouble() / expectedSequence.size

  internal fun recordCorrect(): JamoJudgeState = copy(
    currentIndex = currentIndex + 1,
    correctCount = correctCount + 1,
  )

  internal fun recordError(): JamoJudgeState = copy(errorCount = errorCount + 1)

  companion object {
    fun forTarget(target: String): JamoJudgeState = JamoJudgeState(
      target = target,
      expectedSequence = JamoDecomposer.keySequenceFor(target),
      currentIndex = 0,
      correctCount = 0,
      errorCount = 0,
    )
  }
}

sealed interface JamoJudgeResult {
  data class Correct(val completed: Boolean) : JamoJudgeResult
  data class Incorrect(val expected: Char) : JamoJudgeResult
  data object AlreadyComplete : JamoJudgeResult
}

data class JamoJudgeEvaluation(
  val state: JamoJudgeState,
  val result: JamoJudgeResult,
)

object JamoSequenceJudge {
  /** Compares one logical jamo event. Shift+jamo arrives as its resulting jamo and counts once. */
  fun evaluate(input: Char, state: JamoJudgeState): JamoJudgeEvaluation {
    val expected = state.expectedNext
      ?: return JamoJudgeEvaluation(state, JamoJudgeResult.AlreadyComplete)
    if (input != expected) {
      return JamoJudgeEvaluation(
        state.recordError(),
        JamoJudgeResult.Incorrect(expected),
      )
    }
    val next = state.recordCorrect()
    return JamoJudgeEvaluation(
      next,
      JamoJudgeResult.Correct(completed = next.isComplete),
    )
  }
}
